wpimath/src/main/native/include/frc/DARE.h

```diff
@@ -65,52 +154,131 @@ J = Σ [uₖ] [0 R][uₖ] ΔT
    k=0
 @endverbatim
 
+This internal function skips expensive precondition checks for increased
+performance. The solver may hang if any of the following occur:
+<ul>
+  <li>Q − NR⁻¹Nᵀ isn't symmetric positive semidefinite</li>
+  <li>R isn't symmetric positive definite</li>
+  <li>The (A, B) pair isn't stabilizable</li>
+  <li>The (A, C) pair where Q = CᵀC isn't detectable</li>
+</ul>
+Only use this function if you're sure the preconditions are met.
+
+@tparam States Number of states.
+@tparam Inputs Number of inputs.
 @param A The system matrix.
 @param B The input matrix.
 @param Q The state cost matrix.
 @param R The input cost matrix.
 @param N The state-input cross cost matrix.
-@throws std::invalid_argument if Q − NR⁻¹Nᵀ isn't symmetric positive
-  semidefinite.
-@throws std::invalid_argument if R isn't symmetric positive definite.
-@throws std::invalid_argument if the (A, B) pair isn't stabilizable.
-@throws std::invalid_argument if the (A, C) pair where Q = CᵀC isn't detectable.
 */
-WPILIB_DLLEXPORT
-Eigen::MatrixXd DARE(const Eigen::Ref<const Eigen::MatrixXd>& A,
-                     const Eigen::Ref<const Eigen::MatrixXd>& B,
-                     const Eigen::Ref<const Eigen::MatrixXd>& Q,
-                     const Eigen::Ref<const Eigen::MatrixXd>& R,
-                     const Eigen::Ref<const Eigen::MatrixXd>& N);
+template <int States, int Inputs>
+Eigen::Matrix<double, States, States> DARE(
+    const Eigen::Matrix<double, States, States>& A,
+    const Eigen::Matrix<double, States, Inputs>& B,
+    const Eigen::Matrix<double, States, States>& Q,
+    const Eigen::Matrix<double, Inputs, Inputs>& R,
+    const Eigen::Matrix<double, Inputs, Inputs>& N) {
+  auto R_llt = R.llt();
+  if (R_llt.info() != Eigen::Success) {
+    std::string msg = fmt::format("R isn't positive definite!\n\nR =\n{}\n", R);
+    throw std::invalid_argument(msg);
+  }
```

Comment:
For here and the ABQR overload, does the existence of this check match the doc comments? (Additionally, this check will be performed again when it calls the ABQR overload, but I have no idea how expensive it is or if removing it would make sense conceptually)

Reply:
I'll look into refactoring things to avoid the second LLT decomposition.